File: lab2-functionalprimer/python3/func_as_arg.py
def listFunc(a, b):
    return list(range(a, b+1)) #Create list of ints from 1 to 5, Haskell equivalent [1..5]

def applicatorFunc(inpFunc, s):
    if s=='s':
        return sum(inpFunc())
    else:
        return sum(inpFunc())/len(inpFunc())

result = applicatorFunc(lambda: listFunc(1, 5), 's')
print(result)

File: lab2-functionalprimer/haskell/func_as_arg.hs
-- create inpFunc with a range [a..b]
inpFunc :: [Int]
inpFunc = [1..5]

-- Define applicatorFunc to accept any list of integers and a flag 's' for sum, 'a' for average
applicatorFunc :: [Int] -> Char -> Double
applicatorFunc inpFunc flag
  | flag == 's' = fromIntegral $ sum inpFunc
  | flag == 'a' = fromIntegral (sum inpFunc) / fromIntegral (length inpFunc)
  | otherwise   = error "Invalid flag"

main :: IO ()
main = do
  let result = applicatorFunc inpFunc 'a' -- Call applicatorFunc with inpFunc and 'a' as args
  putStrLn $ case resultType of
    's' -> "Sum = " ++ show result
    'a' -> "Average = " ++ show result
    _   -> "Invalid result type"

-- Define the type of result you want ('s' for sum, 'a' for average)
resultType :: Char
resultType = 'a'
    